lib/improver/spotdata/site_data.py

```diff
@@ -0,0 +1,302 @@
+
+    def json_file(self, file_path):
+        """
+        Use a json file created from the current SSPS system.
+
+        Args:
+        -----
+        file_path : string
+            Path to json file containing SpotData site information.
+
+        Returns:
+        --------
+        sites : dict
+            Dictionary of site data.
+
+        Raises:
+        -------
+        KeyError : If longitude or latitude information is not found.
+
+        """
+        data_file = open(file_path, 'r')
+        site_data = json.load(data_file)
+        data_file.close()
+
+        latitude_entries = [i_site for (i_site, site) in enumerate(site_data)
+                            if 'latitude' in site.keys()]
+        longitude_entries = [i_site for (i_site, site) in enumerate(site_data)
+                             if 'longitude' in site.keys()]
+
+        if not latitude_entries or not longitude_entries:
+            raise KeyError('longitude and latitude must be defined for '
+                           'site in site_data file')
+
+        # Only proceed with sites for which both latitude and longitude are
+        # defined.
+        if latitude_entries != longitude_entries:
+            valid_entries = set(latitude_entries).intersection(
+                longitude_entries)
+            warnings.warn(
+                'Unequal no. of latitudes({}) and longitudes ({}). '
+                'Proceeding with {} complete site definitions.'.format(
+                    len(latitude_entries), len(longitude_entries),
+                    len(valid_entries)))
+        else:
+            valid_entries = latitude_entries
+
+        site_data = [site_data[i_site] for i_site in valid_entries]
+
+        self.latitudes = np.array([site['latitude'] for site in site_data])
+        self.longitudes = np.array([site['longitude'] for site in site_data])
+
+        n_sites = len(self.latitudes)
+
+        # If altitudes are unset use np.nan to indicate that they are at the
+        # altitude of their neighbouring grid point.
+        self.altitudes = np.full(n_sites, np.nan)
+        for i_site, site in enumerate(site_data):
+            if 'altitude' in site.keys():
+                self.altitudes[i_site] = site['altitude']
+
+        # If no site ids are provided, use index of read order.
+        self.site_ids = np.full(n_sites, np.nan)
+        for i_site, site in enumerate(site_data):
+            if 'bestdata_id' in site.keys():
+                self.site_ids[i_site] = site['bestdata_id']
+        # If unset use key LATLONALTI
+        # e.g. lat=50, lon=10, altitude=0 --> 0500100000
+        unset_indices, = np.where(np.isnan(self.site_ids))
+        for i_site in unset_indices:
+            self.site_ids[i_site] = '{:03d}{:03d}{:04d}'.format(
+                int(self.latitudes[i_site]+180),
+                int(self.longitudes[i_site]+180),
+                int(np.nanmax([self.altitudes[i_site], 0.])))
+
```

Comment:
This change is better than just a random number but it will not give a unique number for a particular lat, lon and altitude and there is no check that the value created isn't an existing site_id. It would be better to have no site_id set that have the same_id for different sites. Although recording the bestdata_id is useful we need to consider whether we still need a site_id or whether we will always derive it from lat, lon and altitude or if there is an easy way of getting a unique number from the full lat, lon and altitude. Recording whether the site is associated with an observing site (WMO, ICAO) would be much more important